lib/Conversion/TritonToLinalgExperimental/CollapseShape.cpp

```diff
@@ -0,0 +1,418 @@
+                               PatternRewriter &rewriter) const {
+    auto loc = op.getLoc();
+    auto result = op.getOutputs()[0];
+    auto resultType = mlir::dyn_cast_or_null<MemRefType>(result.getType());
+    if (!resultType)
+      return failure();
+    auto rank = resultType.getRank();
+    if (rank <= 1 ||
+        dyn_cast_or_null<StridedLayoutAttr>(resultType.getLayout())) {
+      return failure();
+    }
+    SmallVector<ReassociationExprs> reassociationMap;
+    reassociationMap.push_back({});
+    for (unsigned i = 0; i < rank; ++i) {
+      reassociationMap[0].push_back(rewriter.getAffineDimExpr(i));
+    }
+    auto elementType = resultType.getElementType();
+
+    auto output = rewriter.create<memref::CollapseShapeOp>(
+        loc,
+        MemRefType::get(llvm::ArrayRef<int64_t>{resultType.getNumElements()},
+                        elementType),
+        result, reassociationMap);
+    op.getOutputsMutable()[0].set(output.getResult());
+    return success();
+  }
+
+  LogicalResult collapseTensor(linalg::FillOp op,
+                               PatternRewriter &rewriter) const {
+    auto loc = op.getLoc();
+    auto result = op.getResult(0);
+    auto resultType =
+        mlir::dyn_cast_or_null<RankedTensorType>(result.getType());
+    if (!resultType)
+      return failure();
+    auto rank = resultType.getRank();
+    if (rank <= 1) {
+      return failure();
+    }
+    auto elementType = resultType.getElementType();
+
+    SmallVector<ReassociationExprs> reassociationMap;
+    reassociationMap.push_back({});
+    for (unsigned i = 0; i < rank; ++i) {
+      reassociationMap[0].push_back(rewriter.getAffineDimExpr(i));
+    }
+
+    auto init = rewriter.create<tensor::CollapseShapeOp>(
+        loc, RankedTensorType::get({resultType.getNumElements()}, elementType),
+        op.getOutputs()[0], reassociationMap);
+    auto fillOp =
+        rewriter.create<linalg::FillOp>(loc, op.getInputs(), ValueRange{init});
+
+    auto expandOp = rewriter.create<tensor::ExpandShapeOp>(
+        loc, result.getType(), fillOp.getResult(0), reassociationMap);
+
+    rewriter.replaceOp(op, expandOp.getResult());
+    return success();
+  }
+
+  LogicalResult matchAndRewrite(linalg::FillOp op,
+                                PatternRewriter &rewriter) const override {
+    assert(op->getNumResults() <= 1 && "code assumes single result!");
+    if (op->getNumResults() == 1) {
+      return collapseTensor(op, rewriter);
+    } else if (op->getNumResults() == 0) {
+      return collapseMemRef(op, rewriter);
+    }
+    return failure();
+  }
+};
+
+// This pattern collapses a `linalg.transpose` operation that transposes a
+// tensor into a `tensor.collapse_shape` operation that collapses the tensor
+// to a smaller shape. This is useful for optimizing the performance of tensor
+// operations that involve transposing large tensors.
+struct CollapseTranspose : public OpRewritePattern<linalg::TransposeOp> {
```

Comment:
i think it will be worth pasting some snippets of the IR here so it's easier to follow (same for the other patterns)